МФТИ ФИВТ: Курс Машинное Обучение (осень, 2016), Арсений Ашуха, ars.ashuha@gmail.com

<h1 align="center">Organization Info</h1> 

**Дополнительный материал для выполнения дз**:
- Hastie, The Elements of Statistical Learning, https://goo.gl/k3wfEU
    - 2.9 Model Selection and the Bias–Variance Tradeoff 
    - 15 Random Forests
- Соколов, Семинары по композиционным методам, https://goo.gl/sn8RyJ
- Andrew Ng, Bias vs. Variance, https://goo.gl/1ISZ6Y

**Оформление дз**: 
- Присылайте выполненное задание на почту ``ml.course.mipt@gmail.com``
- Укажите тему письма в следующем формате ``ML2016_fall <номер_группы> <фамилия>``, к примеру -- ``ML2016_fall 401 ivanov``
- Выполненное дз сохраните в файл ``<фамилия>_<группа>_task<номер>.ipnb``, к примеру -- ``ivanov_401_task1.ipnb``

**Вопросы**:
- Присылайте вопросы на почту ``ml.course.mipt@gmail.com``
- Укажите тему письма в следующем формате ``ML2016_fall Question <Содержание вопроса>``

--------
- **PS1**: Мы используем автоматические фильтры, и просто не найдем ваше дз, если вы не аккуратно его подпишите.
- **PS2**: Напоминаем, что дедлайны жесткие, письма пришедшие после автоматически удаляются =( чтобы соблазна не было 

<h1 align="center">Check Questions</h1> 

Ответе на вопросы своими словами (загугленный материал надо пересказать), ответ обоснуйте (напишите и ОБЪЯСНИТЕ формулки если потребуется), если не выходит, то вернитесь к лекции дополнительным материалам:

__Возможно ответы на некоторые вопросы не будут полностью соответствовать теории с лекций/семинаров, т.к. я готовился по теории в онлайн-курсе от МФТИ и Яндекса на Coursera__

**Вопрос 1**: Какие формулы у шума, смещения, разброса? Какой смысл у этих компонент?

Ошибку, которая возникает при применении алгоритма к новым данным, можно разложить на 3 компоненты: шум, смещение и разброс. В случае среднеквадратичной ошибки ($y(x)$ --- истинный ответ на объекте $x$, $a(x)$ --- предсказание алгоритма на объекте $x$):  
<font color='blue'>
$\mathbb E((y(x) - a(x))^2) = \mathbb E(y(x)^2 - 2\cdot y(x)\cdot a(x) + a(x)^2) = $
$\mathbb E(a(x)^2) + \mathbb E(y(x)^2) - 2\mathbb E(a(x)\cdot y(x)) = $
$(\mathbb E(a(x)^2) - \mathbb E^2(a(x)) + \mathbb E^2(a(x))) + (\mathbb E(y(x)^2) - \mathbb E^2(y(x)) + \mathbb E^2(y(x))) - 2\mathbb E(a(x)y(x)) = $
$ \mathbb D(a(x)) + \mathbb D(y(x)) + \mathbb E^2(a(x)) + \mathbb E^2(y(x)) - 2\mathbb E(a(x))\mathbb E(y(x))) = $
$ \mathbb D(a(x)) + \mathbb D(y(x)) + (\mathbb E(a(x)) - \mathbb E(y(x)))^2 = $
$ \mathbb D(a(x)) + \mathbb D(y(x)) + \mathbb E^2(a(x) - y(x))$
</font>


_Шум_. Данные в задаче могут быть зашумлены настолько, что на них в принципе невозможно получить нулевую ошибку и что даже идеальная модель будет ошибаться на новых данных. Шум показывает, _насколько_ сильно идеальная модель ошибается на этой задаче. При этом шум невозможно улучшить, т.к. это характеристика самих данных.  
Формула: $\mathbb D(y(x))$.

Далее рассмотрим набор алгоритмов $\{a_i(x)\}_{i=1}^{n}$, построенных на основе разных поднаборов одного и того же набора данных (воспринимаем исходную обучающую выборку как случайную величину с некоторым распределением и генерируем серию выборок из одного и того же распределения). Пусть $\hat{a}(x)$ --- алгоритм, возвращающий для каждого объекта $x$ среднее значение предсказаний алгоритмов $a_i(x)$, а $a_{*}(x)$ --- идеальный алгоритм, решающий нашу задачу.

_Смещение_. По сути, это среднее отклонение предсказаний "среднего" алгоритма $\hat{a}(x)$ от предсказаний идеального алгоритма $a_*(x)$. Показывает, достаточно ли сложно выбранное семейство алгоритмов, чтобы приблизить реальную зависимость.  
Формула: $\mathbb E(a(x) - y(x))$.

_Разброс_. По сути, это дисперсия предсказаний алгоритмов $\{a_i(x)\}_{i=1}^{n}$. Показывает, насколько сильно алгоритм зависит от небольших изменений обучающей выборки.  
Формула: $\mathbb D(a(x))$.

**Вопрос 2**: Приведите пример семейства с маленьким смещением и большим разбросом. Приведите пример семейства с большим смещением и маленьким разбросом.

_Решающие деревья_ --- пример семейства моделей с низким смещением и большим разбросом. Это семейство достаточно сложно, чтобы восстанавливать непростые зависимости, но при этом малейшее изменение обучающей выборки ведет к построению другого дерева.

_Линейные алгоритмы_ --- пример семейства моделей с большим смещением и малым разбросом. Это семейство весьма просто и (по умолчанию, т.е. без применения спрямляющих пространств) не в состоянии восстанавливать сложные (нелинейные) зависимости. Однако при этом малые изменения обучающей выборки практически не влияют на итоговую модель.

**Вопрос 3**: Как сгенерировать подвыборку с помощью бутстрапа?

Пусть исходная выборка выглядит так: $X=\{x_i\}_{i=1}^{l}$.
Тогда метод bootstrap строит подвыборку размера $l$ (того же, что и исходная выборка), путем последовательного выбора с возвращением случайного объекта исходной выборки. Таким образом, в bootstrap-выборку некоторые объекты исходной выборки могут попасть несколько раз, а некоторые, в таком случае, не попадут вовсе. 

**Вопрос 4**: Что такое бэггинг?

Подход, который используется при обучении базовых алгоритмов в построении какой-то композиции. Он заключается в том, что каждый базовый алгоритм обучается не на всей исходной выборке, а на неком наборе ее объектов, полученном путем последовательного выбора с возвращением случайного объекта исходной выборки. Размер подвыборки --- это гиперпараметр. Цель применения бэггинга --- понизить корреляцию между базовыми алгоритмами (сделать их максимально независимыми), т.к. чем более независимы базовые алгоритмы, тем сильнее понижается разброс у их композиции (например, у усреднения базовых алгоритмов), а значит и суммарная ошибка алгоритма на новых данных оказывается меньше.

**Вопрос 5**:  Как соотносятся смещение и разброс композиции, построенной с помощью бэггинга, со смещением и разбросом одного базового алгоритма?

В моем ответе предполагается, что композиция --- это усреднение базовых алгоритмов.
Бэггинг понижает корреляцию между базовыми алгоритмами за счет рандомизации обучающей выборки.
Смещение базового алгоритма и смещение композиции совпадают, исходя из общего смысла смещения, данного в ответе на вопрос 1. А вот разброс при усреднении независимых алгоритмов, в свою очередь, уменьшается в N раз, где N --- количество базовых алгоритмов: $\newcommand{\cov}{\operatorname{cov}} \mathbb D(a(x)) = \mathbb D(\frac {1}{N} \sum\limits_{i = 1}^N a_i(x)) = \sum\limits_{i=1}^N \frac {1}{N^2} \mathbb D(a_i(x)) + 2\sum\limits_{1\le i < j \le N} \frac{1}{N^2}\cdot \cov(a_i, a_j) = $  
/фиксируем $k\in \{1, ..., N\}$/ $ = \sum\limits_{i=1}^N \frac {1}{N^2} \mathbb D(a_k(x)) + 2\sum\limits_{1\le i < j \le N} \frac{1}{N^2}\cdot \cov(a_i, a_j) = $
$ \frac {1}{N} \mathbb D(a_k(x)) + 2\sum\limits_{1\le i < j \le N} \frac{1}{N^2}\cdot \cov(a_i, a_j) $.

Из полученной формулы видно, что если алгоритмы $\{a_i\}_{i=1}^N$ будут независимыми, то $\forall i, j:\ i\ne j: cov(a_i, a_j) = 0$, и разброс уменьшается в $N$ раз.

**Вопрос 6**: Как обучается случайный лес? В чем отличия от обычной процедуры построения решающих деревьев?

Случайный лес является композицией базовых алгоритмов --- решающих деревьев, поэтому его обучение сводится к обучению каждого из решающих деревьев на выборке, полученной методом bootstrap из исходной, и их последующему усреднению. Сам процесс построения базового алгоритма рандомизирован:
- пусть в вершину $k$ пришла выборка $X_k$, хотим разбить эту вершину на две и запустить процесс для каждого из двух новых поддеревьев (продолжаем процесс, пока не выполнится критерий останова);
- $Q(X_k, j, t)$ --- критерий ошибки условия $[x^j \le t]$;
- хотим найти наилучшие (с точки зрения минимизации $Q$) параметры разбиения $j$ (индекс признака, по которому разбиваем) и $t$ (пороговое значение признака);
- ищем $t$ среди всех возможных порогов, как и раньше;
- а вот $j$ будем выбирать не из всех признаков, а из некоторого их поднабора длины $q$.

Таким образом, отличие заключается в том, что при построении случайного леса в выборе оптимального разбиения вершины используется случайный поднабор _признаков_ размера $q$. Этот поднабор фиксируется не для всего дерева (как в методе случайных подпространств), а для каждой вершины генерируется заново. Тем самым базовые деревья получаются более независимыми, чем в методе случайных подпространств. Размер поднабора $q$ --- гиперпараметр, который нередко полагают равным $\sqrt d$ в задачах классификации и равным $\frac {d}{3}$ --- в задачах регрессии (где $d$ --- количество признаков).

**Вопрос 7**: Почему хорошими базовыми алгоритмами для бэггинга являются именно деревья?

Потому что у решающих деревьев низкое смещение, они способны восстанавливать достаточно сложные зависимости. Их существенный минус - это большой разброс, но именно бэггинг помогает справиться с этим минусом. В итоге, видим, что совместное использование бэггинга и решающих деревьев теоретически должно быть удачным.

**Вопрос 8**: Как оценить качество случайного леса с помощью out-of-bag-процедуры?

При построении случайного леса каждый базовый алгоритм обучается на случайной подвыборке, получаемой по методу bootstrap на основе исходной обучающей выборки $X$ (размера $l$). Тем самым, можно заметить, что в большинстве случаев для каждого объекта $x_i$ найдется набор деревьев, в обучающей выборке которых не было объекта $x_i$. Но тогда можно оценивать качество случайного леса на объекте $x_i$, не выделяя часть объектов в тестовую выборку.
Пусть случайный лес строится как композиция $N$ базовых алгоритмов, которые обучались на выборках $\{X_i\}_{i=1}^N$. Тогда качество (ошибку) алгоритма можно оценить так:
$\sum\limits_{i = 1}^l Q(y(x_i), \frac{1}{\sum\limits_{n=1}^N [x_i \notin X_n]}\cdot \sum\limits_{n=1}^N [x_i\notin X_n]\cdot b_n(x_i))$, т.е. суммируем функцию потерь по всем объектам, подавая на вход функции потерь правильный ответ $y(x_i)$ и среднее значение прогнозов на объекте $x_i$ тех базовых алгоритмов, которые не обучались на объекте $x_i$.

-----------
PS: Если проверяющий не понял ответ на большинство вопросов, то будет пичалька. Пишите так, чтобы можно было разобраться. 

<h1 align="center">Bagging</h1> 

Известно, что бэггинг плохо работает, если в качестве базовых классификаторов взять knn. Попробуем понять причины на простом примере.

Пусть дана выборка $X^l$ из $l$ объектов с ответами из множества $Y = \{−1, +1\}$. Будем рассматривать классификатор одного ближайшего соседа в качестве базового алгоритма. Построим с помощью бэггинга композицию длины $N$:

$$a_N(x) = sign(\sum_{n=1}^{N} b_n(x))$$

Оцените вероятность того, что ответ композиции на произвольном объекте x будет
отличаться от ответа одного классификатора ближайшего соседа, обученного по всей
выборке. Покажите, что эта вероятность стремится к нулю при N → ∞.

-----
Зафиксируем объект $x$ и его ближайшего соседа $y$. При бутстрапе в каждой выборке используются ~63.2% объектов исходной выборки. Чтобы ответ композиции отличался от метки класса объекта $y$, необходимо чтобы $x$ попал в обучающие выборки менее чем половины деревьев.  
$P(x$ попал в $X_i) \approx 0.632$.
Тогда $P($ ответ композиции отличается от ответа 1NN $) = P(x$ попал в менее чем половину выборок $ \le (1-0.632)^\frac{N}{2}$.

<h1 align="center">Bagging Implementation</h1> 

Реализуйте беггинг.

In [1]:
import warnings

import numpy as np
from copy import deepcopy
from sklearn.base import ClassifierMixin, BaseEstimator
from scipy import stats

warnings.simplefilter("ignore")


class BaggingClassifier(BaseEstimator, ClassifierMixin):
    def __init__(self, base_estimator, n_estimators, items_rate=1.0, features_rate=1.0):
        """
        Parameters
        ----------
        base_estimator: sklearn.Classifier
            Базовый алгоритм, который можно обучить (есть метод fit).
            Для обучение композиции нужно много таких, можэно получить с помощю copy.deepcopy

        n_estimators: int
            Число алгоритмов в композиции

        items_rate: float > 0
            Доля объектов из трейна, на которой будет обучаться каждый базовый алгоритм

        features_rate: float > 0
            Доля фичей, на которой будет обучаться и применяться каждый базовый алгоритм
        """
        self.base_estimator = base_estimator
        self.n_estimators = n_estimators
        self.items_rate = items_rate
        self.features_rate = features_rate

    def fit(self, X, y):
        """
        Метод должен обучить композицию алгоритмов, используя X, y как обучающую выборку.
        Не забудте реализорвать функционал выбора случайных объектов и фичей.

        Parameters
        ----------
        X: 2d np.array
        y: 1d np.array
        """

        # Тут храните обеченные базовые алгоритмы
        self.estimators = []

        # Тут храните фичи для каждого алгоритма
        self.features_idx = []

        for i in range(self.n_estimators):
            estimator = deepcopy(self.base_estimator)
            # =======================================
            # Обучите базовые алгоритмы
            # =======================================
            
            # generating training sample (object can be repeated, features cannot be repeated)
            objects_cnt, features_cnt = X.shape
            train_objects_cnt = max(int(self.items_rate * objects_cnt), 1)
            train_features_cnt = max(int(self.features_rate * features_cnt), 1)
            objects_idx = np.random.randint(0, high=objects_cnt, size=train_objects_cnt)
            features_idx = np.sort(np.random.permutation(np.arange(features_cnt))[:train_features_cnt])
            self.features_idx.append(features_idx)
            X_train = X[objects_idx, :][:, features_idx]
            y_train = y[objects_idx]

            # training base estimator
            estimator.fit(X_train, y_train)
            self.estimators.append(estimator)
            
        return self

    def predict(self, X):
        """

        Parameters
        ----------
        X: 2d np.array матрица объекты признаки на которых нужно сказать ответ

        Returns
        -------
        y_pred: 1d np.array, Вектор классов для каждого объекта
        """
        # Ввиду того, что автор задания сам не определился с тем, вектор чего (вероятностей или меток классов) 
        # я все-таки должен возвращать, я буду возвращать вектор меток классов.
        
        predictions = [] # Храните тут ответы каждого базового алгоритма 
        
        for estimator, features_idx in zip(self.estimators, self.features_idx):
            # =======================================
            # Получите ответы (метки классов) от всех базовых алгоритмов
            # ======================================
            predictions.append(estimator.predict(X[:, features_idx]))
        # =======================================
        # Усредните предсказания, полученные от базовых алгоритмов
        # =======================================
        predictions = np.array(predictions)
        y_pred = np.reshape(stats.mode(predictions, axis=0)[0], (-1, ))
        
        return y_pred
  
# # My small debug test
# from sklearn.tree import DecisionTreeClassifier
# estimator = BaggingClassifier(DecisionTreeClassifier(), 4, items_rate=0.4, features_rate=0.8)
# X = np.eye(10)
# y = np.arange(10)
# estimator.fit(X, y)
# estimator.predict(np.eye(10))

### Titanic Test

In [2]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import LabelEncoder

titanic = pd.read_csv('./data/train.csv')[['Survived', 'Pclass', 'Sex', 'Age', 'Fare']]

sex_encoder = LabelEncoder()
titanic.Sex = sex_encoder.fit_transform(titanic.Sex)
features = ['Pclass', 'Sex', 'Age', 'Fare']

print titanic.shape
titanic.head()

(891, 5)


Survived  Pclass  Sex   Age     Fare
0         0       3    1  22.0   7.2500
1         1       1    0  38.0  71.2833
2         1       3    0  26.0   7.9250
3         1       1    0  35.0  53.1000
4         0       3    1  35.0   8.0500

In [3]:
X, y = titanic[features].values, titanic.Survived.values
X = np.nan_to_num(X)

X_train_size = 500
X_train, y_train, X_test, y_test = X[:X_train_size], y[:X_train_size], X[X_train_size:], y[X_train_size:]

Нужно обучить свой беггинг на датасете титаник, и посмотреть работает ли он. 

In [4]:
from sklearn.metrics import accuracy_score
from sklearn.tree import DecisionTreeClassifier

In [5]:
# =======================================
# Обучите беггинг над DecisionTreeClassifier с 10 моделями
# =======================================
clf = BaggingClassifier(DecisionTreeClassifier(), 10)
clf.fit(X_train, y_train)
print accuracy_score(clf.predict(X_train), y_train), accuracy_score(clf.predict(X_test), y_test)

0.976 0.787723785166


Проведите эксперименты:
    - Работает-ли беггинг лучше чем просто линейная модель?
    - Какой items_rate и features_rate работает лучше и почему?

In [6]:
# =======================================
# Обучите беггинг над DecisionTreeClassifier с 100 моделями
# =======================================
clf = BaggingClassifier(DecisionTreeClassifier(), 100)
clf.fit(X_train, y_train)
acc = accuracy_score(clf.predict(X_train), y_train), accuracy_score(clf.predict(X_test), y_test)
acc

(0.98799999999999999, 0.78772378516624042)

In [7]:
# =======================================
# Обучите LogsiticRegression 
# =======================================
from sklearn.linear_model import LogisticRegression
clf = LogisticRegression()
clf.fit(X_train, y_train)
acc = accuracy_score(clf.predict(X_train), y_train), accuracy_score(clf.predict(X_test), y_test)
acc

(0.80000000000000004, 0.76982097186700771)

In [8]:
%%time

# Finding good values for items_rate and features_rate
from sklearn.grid_search import GridSearchCV

classifier = BaggingClassifier(DecisionTreeClassifier(), 10)
# print classifier.get_params().keys()
param_grid = {'items_rate': [0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 1],
              'features_rate': [0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 1]}
optimizer = GridSearchCV(estimator=classifier, param_grid=param_grid, cv=3, scoring='accuracy')
optimizer.fit(X_train, y_train)
print optimizer.best_params_
print optimizer.best_score_

{'items_rate': 1, 'features_rate': 0.9}
0.812
CPU times: user 2.6 s, sys: 20 ms, total: 2.62 s
Wall time: 2.62 s


### Adult Test

In [9]:
import pandas as pd
adult = pd.read_csv(
    './data/adult.data', 
    names=[
        "Age", "Workclass", "fnlwgt", "Education", "Education-Num", "Martial Status",
        "Occupation", "Relationship", "Race", "Sex", "Capital Gain", "Capital Loss",
        "Hours per week", "Country", "Target"], 
    header=None, na_values="?")

adult = pd.get_dummies(adult)
adult["Target"] = adult["Target_ >50K"]
X, y = adult[adult.columns[:-3]].values, adult[adult.columns[-1]].values
X_train_size = 20000
X_train, y_train, X_test, y_test = X[:X_train_size], y[:X_train_size], X[X_train_size:], y[X_train_size:]

print adult.shape
adult.head()

(32561, 111)


Age  fnlwgt  Education-Num  Capital Gain  Capital Loss  Hours per week  \
0   39   77516             13          2174             0              40   
1   50   83311             13             0             0              13   
2   38  215646              9             0             0              40   
3   53  234721              7             0             0              40   
4   28  338409             13             0             0              40   

   Workclass_ ?  Workclass_ Federal-gov  Workclass_ Local-gov  \
0           0.0                     0.0                   0.0   
1           0.0                     0.0                   0.0   
2           0.0                     0.0                   0.0   
3           0.0                     0.0                   0.0   
4           0.0                     0.0                   0.0   

   Workclass_ Never-worked   ...    Country_ South  Country_ Taiwan  \
0                      0.0   ...               0.0              0.0   
1                      0.0   ...               0.0              0.0   
2                      0.0   ...               0.0              0.0   
3                      0.0   ...               0.0              0.0   
4                      0.0   ...               0.0              0.0   

   Country_ Thailand  Country_ Trinadad&Tobago  Country_ United-States  \
0                0.0                       0.0                     1.0   
1                0.0                       0.0                     1.0   
2                0.0                       0.0                     1.0   
3                0.0                       0.0                     1.0   
4                0.0                       0.0                     0.0   

   Country_ Vietnam  Country_ Yugoslavia  Target_ <=50K  Target_ >50K  Target  
0               0.0                  0.0            1.0           0.0     0.0  
1               0.0                  0.0            1.0           0.0     0.0  
2               0.0                  0.0            1.0           0.0     0.0  
3               0.0                  0.0            1.0           0.0     0.0  
4               0.0                  0.0            1.0           0.0     0.0  

[5 rows x 111 columns]

In [10]:
%%time
clf = BaggingClassifier(DecisionTreeClassifier(), 10)
clf.fit(X_train, y_train)
acc = accuracy_score(clf.predict(X_train), y_train), accuracy_score(clf.predict(X_test), y_test)
print acc

(0.98765000000000003, 0.84849932330228484)
CPU times: user 5.39 s, sys: 12 ms, total: 5.4 s
Wall time: 5.4 s


In [11]:
%%time
clf = BaggingClassifier(DecisionTreeClassifier(), 100)
clf.fit(X_train, y_train)
acc = accuracy_score(clf.predict(X_train), y_train), accuracy_score(clf.predict(X_test), y_test)
print acc

(1.0, 0.85765464533078573)
CPU times: user 55.9 s, sys: 68 ms, total: 56 s
Wall time: 56 s


In [12]:
%%time
clf = LogisticRegression()
clf.fit(X_train, y_train)
acc = accuracy_score(clf.predict(X_train), y_train), accuracy_score(clf.predict(X_test), y_test)
print acc

(0.79990000000000006, 0.79515962104927951)
CPU times: user 392 ms, sys: 0 ns, total: 392 ms
Wall time: 382 ms


In [13]:
%%time

classifier = BaggingClassifier(DecisionTreeClassifier(), 10)
# print classifier.get_params().keys()
param_grid = {'items_rate': [0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 1],
              'features_rate': [0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 1]}
optimizer = GridSearchCV(estimator=classifier, param_grid=param_grid, cv=3, scoring='accuracy')
optimizer.fit(X_train, y_train)
print optimizer.best_params_
print optimizer.best_score_

{'items_rate': 0.75, 'features_rate': 0.75}
0.85015
CPU times: user 1min 26s, sys: 100 ms, total: 1min 26s
Wall time: 1min 27s


Ответте на вопросы:
    - Работает-ли беггинг лучше чем просто линейная модель?
    - Какой items_rate и features_rate работает лучше и почему?

__Работает ли бэггинг лучше чем линейная модель?__
В обоих тестах BaggingClassifier показывает лучший результат, чем линейная модель, как на обучающей, так и на тестовой выборках. При этом композиция с 100 базовыми алгоритмами дает лучшие предсказания, чем композиция с 10 базовыми алгоритмами.

__Какой items_rate и features_rate работает лучше и почему?__  
В тесте _Titanic_ оказалось оптимальным генерировать обучающие выборки из всех объектов исходной выборки и при этом для каждого базового алгоритма рассматривать 90% всех признаков.

В тесте _Adult_ оказалось оптимальным при обучении базовых алгоритмов использовать три четверти объектов, и три четверти признаков.

Судя по всему, это объясняется тем, что тест _Titanic_ содержит малое количество объектов (891), что не позволяет алгоритму достаточно качественно обучиться при "выбрасывании" части объектов. В свою очередь, тест _Adult_ содержит много объектов (32561), которых достаточно не только для качественного обучения, но и для реализации бэггинга. Также, возможно, целевая метка сильно зависит от всех признаков, поэтому при обучении базовых алгоритмов в обоих тестах используются почти все признаки.

<h1 align="center">Text, Image Classification</h1> 

Дальше в каждом эксперименте нужно: 
- сравниться с линейной моделью ( какую лучше выбрать?=) )
- сделать выбор в пользу одной из моделей
- выбор обосновать, почему одна из моделей хуже а другая лучше
- что такое хуже и лучше
- попробуйте беггинг над деревьями и линейными моделями 
- почему работает или не работает, какие особенности данных на это влияют

### Text classification

In [14]:
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import TfidfVectorizer
newsgroups_train = fetch_20newsgroups(subset='train')
newsgroups_test = fetch_20newsgroups(subset='test')

In [15]:
vectorizer = TfidfVectorizer()
X_train, y_train = vectorizer.fit_transform(newsgroups_train.data), newsgroups_train.target
X_test,  y_test  = vectorizer.transform(newsgroups_test.data), newsgroups_test.target

In [16]:
print X_train.shape
print y_train.shape
print y_train[:5]

(11314, 130107)
(11314,)
[ 7  4  4  1 14]


In [17]:
%%time
# =======================================
# Обучите Линейную модель 
# =======================================

# 1. Обучим логистическую регрессию с l1-регуляризацией (Lasso).
from sklearn.linear_model import LogisticRegression

clf = LogisticRegression(penalty='l1')
clf.fit(X_train, y_train)
print accuracy_score(clf.predict(X_train), y_train), accuracy_score(clf.predict(X_test), y_test)
features_left = np.count_nonzero(clf.coef_)
print 'Features deleted: {} / {}'.format(X_train.shape[1] - features_left, X_train.shape[1])

0.868658299452 0.752124269782
Features deleted: 128499 / 130107
CPU times: user 19.1 s, sys: 28 ms, total: 19.1 s
Wall time: 19.1 s


In [18]:
%%time

# 2. Обучим логистическую регрессию с l2-регуляризацией (Ridge).
clf = LogisticRegression(penalty='l2')
clf.fit(X_train, y_train)
print accuracy_score(clf.predict(X_train), y_train), accuracy_score(clf.predict(X_test), y_test)

0.969860350009 0.827934147637
CPU times: user 45.1 s, sys: 780 ms, total: 45.9 s
Wall time: 23.1 s


In [19]:
%%time

# 3. Обучим классификатор на основе стохастического градиентного спуска (SGD)
from sklearn.linear_model import SGDClassifier
clf = SGDClassifier()
clf.fit(X_train, y_train)
print accuracy_score(clf.predict(X_train), y_train), accuracy_score(clf.predict(X_test), y_test)

0.993636202934 0.854753053638
CPU times: user 3.61 s, sys: 80 ms, total: 3.69 s
Wall time: 1.87 s


Как можно видеть, классификатор на основе стохастического градиентного спуска дает наилучший результат. Честно говоря, я не знаю, как это объяснить. Я ожидал, что наилучший результат будет у логистической регрессии с l1-регуляризацией (Lasso), т.к. в данных огромное число признаков, многие из которых наверняка избыточны и могли бы быть обнулены с помощью l1-регуляризации, оставляя лишь те, которые действительно влияют на целевые метки классов.

In [20]:
%%time
# =======================================
# Обучите беггинг над DecisionTreeClassifier
# =======================================

clf = BaggingClassifier(DecisionTreeClassifier(), 10)
clf.fit(X_train, y_train)
print accuracy_score(clf.predict(X_train), y_train), accuracy_score(clf.predict(X_test), y_test)

0.989482057628 0.631970260223
CPU times: user 6min 27s, sys: 824 ms, total: 6min 28s
Wall time: 6min 29s


### Image classification

In [21]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [22]:
from utils import load_cifar10
X_train, y_train, X_val, y_val, X_test, y_test = load_cifar10('./data/cifar10')[0]

In [23]:
X_train, X_test = X_train.reshape(X_train.shape[0], -1), X_test.reshape(X_test.shape[0], -1)
print X_train.shape
print X_test.shape

(50000, 3072)
(10000, 3072)


In [24]:
%%time
# Обучим классификатор на основе стохастического градиентного спуска (SGD), т.к. 
# он способен отработать за адекватное время на данных такого масштаба
from sklearn.linear_model import SGDClassifier
clf = SGDClassifier()
clf.fit(X_train, y_train)
print accuracy_score(clf.predict(X_train), y_train), accuracy_score(clf.predict(X_test), y_test)

0.21284 0.1964
CPU times: user 33.8 s, sys: 16 ms, total: 33.8 s
Wall time: 33.4 s


In [22]:
%%time
# =======================================
# Обучите беггинг над DecisionTreeClassifier
# =======================================
clf = BaggingClassifier(DecisionTreeClassifier(), 10)
clf.fit(X_train, y_train)
print accuracy_score(clf.predict(X_train), y_train), accuracy_score(clf.predict(X_test), y_test)

0.99362 0.375
CPU times: user 51min 33s, sys: 19.5 s, total: 51min 53s
Wall time: 52min


<h1 align="center">Random Forest Feature Impotance</h1> 

Опишите как вычисляется важность фичей в дереве, можите изучить как работает  feature\_importances_ в sklearn.

---

При построении решающего дерева происходят разбиения вершин. При разбиении вершины на две критерий Джини двух новых вершин меньше критерия Джини родительской вершины. Если зафиксировать конкретный признак $k$ и просуммировать изменение критерия Джини по всем вершинам, в которых происходило разбиение по признаку $k$, то получим важность признака $k$. Т.е., если в вершине $v$ произошло разбиение по признаку $k$, то в сумму необходимо добавить разность между критерием Джини в вершине $v$ и критерием Джини в ее предке.

Почитайте Feature Impotance для Adult и Titanic (используйте полный датасет), ПРОИНТЕРПРЕТИРУЙТЕ резульататы.

### Adult

In [33]:
X, y = adult[adult.columns[:-3]].values, adult[adult.columns[-1]].values
X_train, y_train, X_test, y_test = X[:20000], y[:20000], X[20000:], y[20000:]

In [34]:
clf = BaggingClassifier(base_estimator=DecisionTreeClassifier(), n_estimators=100, 
                        items_rate=1, features_rate=1).fit(X_train, y_train)

In [35]:
# =======================================
# Посчитайте feature_importances для clf
# =======================================
def feature_importances(clf, X):
    res = np.zeros((X.shape[1], ))
    for estimator in clf.estimators:
        res += estimator.feature_importances_
    return res

importances = feature_importances(clf, X_train)
print importances

[  1.13357977e+01   1.68730886e+01   1.09871352e+01   1.04427259e+01
   3.79384541e+00   6.45116928e+00   1.54181183e-01   5.06054523e-01
   6.14855385e-01   0.00000000e+00   1.00528588e+00   6.61686532e-01
   8.80579449e-01   4.57376567e-01   9.30268723e-04   6.35133262e-02
   6.38527635e-02   4.88109627e-02   7.85148325e-03   1.92973781e-02
   4.73700219e-02   4.98730166e-02   1.40705989e-01   2.62921345e-01
   2.62920070e-01   7.02222142e-02   3.38385741e-01   2.64721265e-01
   7.48336709e-03   1.21636191e-01   4.14911092e-01   2.27228718e-01
   5.11526598e-02   1.95479134e+01   6.68555864e-02   1.90228035e-01
   1.11632223e-01   1.08591128e-01   1.41401336e-01   6.18514199e-01
   1.71535651e-03   8.48215931e-01   1.20923750e+00   3.67192940e-01
   3.04704484e-01   5.44535260e-01   4.53637794e-01   8.12604584e-04
   8.82065413e-01   3.48070852e-01   7.89393979e-01   5.58488593e-01
   5.68289244e-01   2.41976785e-01   2.27471090e-01   1.14910350e-01
   1.60610188e-01   1.41586147e-01

In [37]:
unused_features = np.where(importances == 0)[0]
print unused_features
print adult.columns[unused_features]
# Видим, что признаки 9, 81, 91, 104 абсолютно не участвуют в построении деревьев

[  9  81  94 104]
Index([u'Workclass_ Never-worked', u'Country_ Holand-Netherlands',
       u'Country_ Outlying-US(Guam-USVI-etc)', u'Country_ Trinadad&Tobago'],
      dtype='object')


### Titanick

In [28]:
X, y = titanic[features].values, titanic.Survived.values
X = np.nan_to_num(X)
X_train, y_train, X_test, y_test = X[:500], y[:500], X[500:], y[500:]

In [29]:
clf = BaggingClassifier(base_estimator=DecisionTreeClassifier(), n_estimators=100, 
                        items_rate=1, features_rate=1).fit(X_train, y_train)

In [30]:
# =======================================
# Посчитайте feature_importances для clf
# =======================================

In [31]:
importances = feature_importances(clf, X_train)
print importances
print np.where(importances == 0)[0]
# А здесь все признаки информативны

[  8.14927042  33.44345425  24.95172909  33.45554624]
[]
